DMCompiler/DM/Visitors/DMProcBuilder.cs

```diff
@@ -161,8 +172,24 @@ public void ProcessStatementSet(DMASTProcStatementSet statementSet)
         {
             var attribute = statementSet.Attribute.ToLower();
             // TODO deal with "src"
-            if (!DMExpression.TryConstant(_dmObject, _proc, statementSet.Value, out var constant) && attribute != "src") {
-                throw new CompileErrorException(statementSet.Location, $"{attribute} attribute should be a constant");
+            if(attribute == "src") {
+                DMCompiler.UnimplementedWarning(statementSet.Location, "'set src' is unimplemented");
+                return;
+            }
+            if (!DMExpression.TryConstant(_dmObject, _proc, statementSet.Value, out var constant)) { // If this set statement's rhs is not constant
+                bool didError = DMCompiler.Emit(WarningCode.InvalidSetStatement, statementSet.Location, $"'{attribute}' attribute should be a constant");
+                if (didError) // if this is an error
+                    return; // don't do the cursed thing
+                // oh no.
+                if (previousSetStatementValue is null)
+                    throw new CompileErrorException(statementSet.Location, $"'{attribute}' attribute must be a constant"); // FIXME: Manual promotion of errors would be cool here
+                constant = previousSetStatementValue;
+            } else {
+                previousSetStatementValue = constant;
+            }
+            if (statementSet.WasInKeyword) { // check if it was 'set x in y' or whatever                               // (which is illegal for everything except setting src to something)
```

Comment:
Put these comments on their own lines instead of the huge spacing they currently have.
```suggestion
            }
            
            // check if it was 'set x in y' or whatever
            // (which is illegal for everything except setting src to something)
            if (statementSet.WasInKeyword) {
```